// hotel/src/container/SinglePage/AmountBar/AmoutBar.style.js
import styled from 'styled-components';
import { themeGet } from '@styled-system/theme-get';
const AmountBar = styled.div`
  padding: 0px 0px 0px 0px;
  display:flex;
  margin-left:10px; 
  form {
    
    width: 300px;
    padding-bottom: 20px;
    text-align: center;
    padding: 0px;
    display: flex;
  }

  .value-button {
    display: inline-block;
    border: 1px solid #ddd;

    width: 40px;
    height: 35px;
    font-size: 22px;
    text-align: center;
    vertical-align: middle;

    background: #eee;
    -webkit-touch-callout: none;
    -webkit-user-select: none;
    -khtml-user-select: none;
    -moz-user-select: none;
    -ms-user-select: none;
    user-select: none;
  }

  .value-button:hover {
    cursor: pointer;
  }

  form #decrease {
    border-radius: 0px 0 0 0px;
  }

  form #increase {
    border-radius: 0 0px 0px 0;
  }

  /* form #input-wrap {
    margin: 0px;
    padding: 0px;
  } */

  input#number {
    text-align: center;

    font-size: 22px;
    border: 1px solid #ddd;

    margin: 0px;
    width: 50px;
    height: 35px;
  }

  input[type='number']::-webkit-inner-spin-button,
  input[type='number']::-webkit-outer-spin-button {
    -webkit-appearance: none;
    margin: 0;
  }
`;
export default AmountBar;
